### ISSUE TITLE
Can't push multiple images to same repo in parallel

### ISSUE THREAD

**Issue by phemmer** [docker/docker#9132]
Attempting to push multiple images to the same remote in parallel cases one of the push operations to fail. The operation claims "is already in progress", implying that the image is already being pushed, but this is false. The image is not pushed, and pushing again at a later time will end up actually pushing the image.

```
$ cat > Dockerfile <<'EOF'
FROM fedora:20
ADD file /file
RUN dd if=/dev/zero of=zero bs=1M count=10
EOF

$ echo 1 > file

$ docker build -t phemmer/test:1 .

$ echo 2 > file

$ docker build -t phemmer/test:2 .

$ docker push phemmer/test:1 & docker push phemmer/test:2 ; wait
[1] 31920
The push refers to a repository [phemmer/test] (len: 1)
Sending image list
2014/11/12 20:38:02 Error: push phemmer/test is already in progress
[1]  + exit 1     docker push phemmer/test:1
Pushing repository phemmer/test (1 tags)
511136ea3c5a: Image already pushed, skipping 
c69cab00d6ef: Image already pushed, skipping 
88b42ffd1f7c: Image already pushed, skipping 
64dd2ddab00b: Image successfully pushed 
712a4eb2c2eb: Image successfully pushed 
Pushing tag for rev [712a4eb2c2eb] on {https://cdn-registry-1.docker.io/v1/repositories/phemmer/test/tags/2}
```

In this case, the `phemmer/test:2` image successfully pushed, but `phemmer/test:1` failed.

Docker version 1.3.1

**Comment by clkao** [comment]
In any case, we can make push wait for the other process to finish for the moment.

**Comment by bridgetkromhout** [comment]
> In any case, we can make push wait for the other process to finish for the moment.

This would be great for us. Right now, our jenkins builds break because of this; we get "Error: push ourlocalrepo.example:5000/imagename is already in progress" if ourlocalrepo.example:5000/imagename:dev and ourlocalrepo.example:5000/imagename:qa are both being pushed at the same time. It would be nice not to have to put our own `sleep` in.

Thanks in advance!

**Comment by termie** [comment]
+1

**Comment by daniel-bryant-uk** [comment]
+1 - this is also causing us issues with building and pushing via Jenkins

**Comment by mikejihbe** [comment]
This is a severe limitation in any continuous delivery workflow. Multiple tags being built in parallel is the norm, so docker really ought to be able to handle receiving multiple images in parallel. Is this on the roadmap?

**Comment by tal-moshayov** [comment]
this has a negative impact on our CI flows. working with branches is the recommended approach to development, and this limitation kills it. Is it a docker-hub specific limitation?

**Comment by bydga** [comment]
any progress on this? Pushing multiple versions of the same image in parallell is really important feature for CI. We are now bypassing this by putting the version into docker image name (and not using the version tag at all - everything is just "latest").

**Comment by ejholmes** [comment]
:+1: this is a real pain if you're trying to build a build pipeline around docker. Any of of the maintainers able to weigh in on the rationale behind the addition of [this](https://github.com/docker/docker/blob/6e64e0548f30fd9d1a368ac5ea5bde9dec38c162/graph/tags.go#L404-L406)? I'd be happy to attempt to fix, but I'm curious if there was a technical limitation behind adding the constraint.

**Comment by burke** [comment]
We're feeling the pain of this too in our CI build infrastructure. I'm going to dig in this week and try to come up with a solution.

cc @csfrancis @shivnagarajan

**Comment by kylechadha** [comment]
We're facing this as well, has anyone tried using sleep as a workaround? Seems like a logical solution but not sure if there are other ramifications.

cc @micahasmith @pinetrees

**Comment by jimjh** [comment]
Not sure if this has been mentioned, but the most obvious workaround is

```bash
#!/bin/bash
# /usr/local/sbin/docker-push
exec flock -x /var/lock/docker-pool docker push "${@}"
```

```bash
#!/bin/bash
# /usr/local/sbin/docker-pull
exec flock -x /var/lock/docker-pool docker pull "${@}"
```

**Comment by dpgeekzero** [comment]
Nice one @jimjh, I hadn't thought of that one!

**Comment by thaJeztah** [comment]
ping @aaronlehmann just checking if there's progress in this area, following https://github.com/docker/docker/pull/16891#issuecomment-151338707

**Comment by aaronlehmann** [comment]
It's possible to push multiple images in parallel as of #17924. I think this can be closed now.

**Comment by thaJeztah** [comment]
Thanks for looking @aaronlehmann, I'll close this issue, but anyone that's still having issues after 1.10 has been release, please comment here!

**Comment by lhuard1A** [comment]
Hi,

I’ve just made a try of parallel push with docker 1.10 and it seems there are still some bugged use-cases.

Here is my docker setup:
```
$ docker version
Client:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Thu Feb 25 09:29:05 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Thu Feb 25 09:29:05 2016
 OS/Arch:      linux/amd64
```
```
$ docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.10.2
Storage Driver: btrfs
 Build Version: Btrfs v4.4
 Library Version: 101
Execution Driver: native-0.2
Logging Driver: journald
Plugins: 
 Volume: local
 Network: host bridge null
Kernel Version: 4.3.3-3-ARCH
Operating System: Arch Linux
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 7.755 GiB
…
```
(I also observed exactly the same behaviour with the same version of docker running on an Ubuntu using AUFS)

When I try to push in parallel two docker images which have the same name but two different tags and which share common layers, one push succeeds:
```
$ docker push l3n41c/archlinux:2016.03.01
The push refers to a repository [docker.io/l3n41c/archlinux]
a9cccdd3d595: Pushed 
2016.03.01: digest: sha256:645d16a64140f9614f4b2b21393b7f304293ae041269d6a57ac3aadb21b7bf42 size: 1186
```
whereas the other one fails:
```
$ docker push l3n41c/archlinux:2016.03.01-1
The push refers to a repository [docker.io/l3n41c/archlinux]
a9cccdd3d595: Pushed 
unexpected EOF
```

As I launched the two commands in parallel, both were pushing the same layer at the same time (progress bars about pushing `a9cccdd3d595` progressed at the same rate).
But once the layer was pushed, one command succeeded whereas the other one failed.

And when I look at the docker hub, only `l3n41c/archlinux:2016.03.01` is present. `l3n41c/archlinux:2016.03.01-1` is really missing from the docker hub. So, the issue is not just the client exiting abnormally. The image is really missing from the hub.

**Comment by thaJeztah** [comment]
ping @aaronlehmann ˆˆ

**Comment by aaronlehmann** [comment]
Does l3n41c/archlinux:2016.03.01-1 push successfully when pushing it alone? "Unexpected EOF" shows up sometimes with images that were migrated incorrectly by 1.10.0 or 1.10.1. Can you please confirm that the image itself isn't the source of the problem, and this is definitely related to pushing it in parallel?

**Comment by lhuard1A** [comment]
In fact, my reproducer looks like something like this:

```bash
docker tag $my_image my_user/test:1
docker tag $my_image my_user/test:2
docker push my_user/test:1 > push1 2>&1 &
push1_pid=$!
docker push my_user/test:2 > push2 2>&1 &
push2_pid=$!
wait $push1_pid && echo OK || echo Failed
wait $push2_pid && echo OK || echo Failed
```

**Comment by lhuard1A** [comment]
I don’t think it is specific to my image. I reproduced it also with other images and I also reproduced it with some images which were sharing some layers but not all of them.
It’s just my “minimal” reproducer which is on a single-layer image.

**Comment by thaJeztah** [comment]
Added Priority and moved to 1.10.3 milestone, because of the comments above

**Comment by clkao** [comment]
Doesn't this need to be escalated to P0 for a patch release, according to the [release process](https://github.com/docker/docker/blob/master/project/PATCH-RELEASES.md)?

**Comment by thaJeztah** [comment]
@clkao as you can see from this issue, parallel push to the *same* repository not working, was already reported with docker 1.3.3, two years ago, so the fact that it doesn't work is not a *new* issue.

**Comment by skairali** [comment]
@thaJeztah  is this issue solved already?

**Comment by thaJeztah** [comment]
@skairali yes, this was included in the 1.10.3 release, **but** requires a registry 2.3.x; Docker Hub is not yet using 2.3.x, so you can only benefit from this feature if you're on a private repository running that version
